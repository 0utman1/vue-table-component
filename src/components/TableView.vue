<template>
    <div>
        <table>
            <thead>
            <tr>
                <th v-for="column in columns">
                    {{ column.name }}
                </th>
            </tr>
            </thead>
            <tbody>
                <table-row v-for="row in rows" :row="row" :columns="columns" />
            </tbody>
        </table>
    </div>
</template>

<script>
    import TableRow from './TableRow';

    export default {
        components: { TableRow },

        props: {
            rows: {
                required: true,
            },
            sortBy: {
                type: String,
                required: false,
                default: '',
            },
            sortOrder: {
                type: String,
                required: false,
                default: 'desc',
            }
        },

        data: () => ({}),

        created() {
            this.columns = this.$children;
        },

        computed: {},
    };

</script>
