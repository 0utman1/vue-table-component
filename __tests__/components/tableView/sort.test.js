import TableView from '../../../src';
import Vue from 'vue/dist/vue.js';

describe('TableView is sortable', () => {
    Vue.use(TableView);

    it('can sort the data with by a specific column', async () => {
        setDocumentInnerHtml({ sortBy: 'firstName' });

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can sort the data with by another specific column', async () => {
        setDocumentInnerHtml({ sortBy: 'lastName' });

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can sort the data with by a specific column in a specific order', async () => {
        setDocumentInnerHtml({ sortBy: 'firstName', sortOrder: 'desc' });

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });
});

function setDocumentInnerHtml({ sortBy, sortOrder = '' }) {

    document.body.innerHTML = `
            <div id="app">
                <div>
                    <table-view
                        :data="[{ id: 1, firstName: 'Jay', lastName: 'Vleugels' },
                                { id: 2, firstName: 'Wesley', lastName: 'Biets' },
                                { id: 3, firstName: 'Randy', lastName: 'Paret' },
                                { id: 4, firstName: 'Devon', lastName: 'Macharis' }]"
                        sort-by="${sortBy}"
                        sort-order="${sortOrder}"
                    >
                        <table-column for="firstName" label="First name"></table-column>
                        <table-column for="lastName" label="Last name"></table-column>
                    </table-view>
                </div>
            </div>
        `;
}

async function createVm() {
    const vm = new Vue({
        el: '#app',
    });

    await Vue.nextTick(() => {});
}
