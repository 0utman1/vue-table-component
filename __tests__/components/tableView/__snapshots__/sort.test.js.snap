// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sortable tableView can sort the data with by a specific column 1`] = `
"
            <div id=\\"app\\"><div><div><input type=\\"text\\" name=\\"table-view-filter\\"> <!----> <table><thead><tr><th class=\\"sort-\\">
    First name
</th><th>
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>Devon</td><td>Macharis</td><td>Yellow</td></tr><tr><td>Jay</td><td>Vleugels</td><td>Red</td></tr><tr><td>Randy</td><td>Paret</td><td>Green</td></tr><tr><td>Wesley</td><td>Biets</td><td>Blue</td></tr></tbody></table> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;

exports[`Sortable tableView can sort the data with by a specific column in a specific order 1`] = `
"
            <div id=\\"app\\"><div><div><input type=\\"text\\" name=\\"table-view-filter\\"> <!----> <table><thead><tr><th class=\\"sort-desc\\">
    First name
</th><th>
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>Wesley</td><td>Biets</td><td>Blue</td></tr><tr><td>Randy</td><td>Paret</td><td>Green</td></tr><tr><td>Jay</td><td>Vleugels</td><td>Red</td></tr><tr><td>Devon</td><td>Macharis</td><td>Yellow</td></tr></tbody></table> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;

exports[`Sortable tableView can sort the data with by another specific column 1`] = `
"
            <div id=\\"app\\"><div><div><input type=\\"text\\" name=\\"table-view-filter\\"> <!----> <table><thead><tr><th>
    First name
</th><th class=\\"sort-\\">
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>Wesley</td><td>Biets</td><td>Blue</td></tr><tr><td>Devon</td><td>Macharis</td><td>Yellow</td></tr><tr><td>Randy</td><td>Paret</td><td>Green</td></tr><tr><td>Jay</td><td>Vleugels</td><td>Red</td></tr></tbody></table> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;

exports[`Sortable tableView it will change the sort order when clicking the header of the column with the active sort 1`] = `
"
            <div id=\\"app\\"><div><div><input type=\\"text\\" name=\\"table-view-filter\\"> <!----> <table><thead><tr><th class=\\"sort-desc\\">
    First name
</th><th>
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>Wesley</td><td>Biets</td><td>Blue</td></tr><tr><td>Randy</td><td>Paret</td><td>Green</td></tr><tr><td>Jay</td><td>Vleugels</td><td>Red</td></tr><tr><td>Devon</td><td>Macharis</td><td>Yellow</td></tr></tbody></table> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;

exports[`Sortable tableView will not sort data when clicking a non-sortable column header 1`] = `
"
            <div id=\\"app\\"><div><div><input type=\\"text\\" name=\\"table-view-filter\\"> <!----> <table><thead><tr><th class=\\"sort-\\">
    First name
</th><th>
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>Devon</td><td>Macharis</td><td>Yellow</td></tr><tr><td>Jay</td><td>Vleugels</td><td>Red</td></tr><tr><td>Randy</td><td>Paret</td><td>Green</td></tr><tr><td>Wesley</td><td>Biets</td><td>Blue</td></tr></tbody></table> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;

exports[`Sortable tableView will not sort on a column that is not sortable 1`] = `
"
            <div id=\\"app\\"><div><div><input type=\\"text\\" name=\\"table-view-filter\\"> <!----> <table><thead><tr><th>
    First name
</th><th>
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>Devon</td><td>Macharis</td><td>Yellow</td></tr><tr><td>Jay</td><td>Vleugels</td><td>Red</td></tr><tr><td>Randy</td><td>Paret</td><td>Green</td></tr><tr><td>Wesley</td><td>Biets</td><td>Blue</td></tr></tbody></table> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;

exports[`Sortable tableView will sort the data descendingly if the header of of column without the active sort is clicked 1`] = `
"
            <div id=\\"app\\"><div><div><input type=\\"text\\" name=\\"table-view-filter\\"> <!----> <table><thead><tr><th class=\\"\\">
    First name
</th><th class=\\"sort-desc\\">
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>Jay</td><td>Vleugels</td><td>Red</td></tr><tr><td>Randy</td><td>Paret</td><td>Green</td></tr><tr><td>Devon</td><td>Macharis</td><td>Yellow</td></tr><tr><td>Wesley</td><td>Biets</td><td>Blue</td></tr></tbody></table> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;
